### ISSUE TITLE
Wrong? Conversion with numbers in the string?

### ISSUE THREAD

**Issue by RoNoLo** [gustavoguichard/string-ts#97]
Hi, 

I am not sure if I do not know some kind of standard of camelcase to snakecase conversation, but this is the second library which transforms strings so, that not the original string is produced if using a double convert like toCamel(toSnake("string")).

As far as I understand the converting of camelcase and snakecase and vis a vis the original string should always be the result, if I double convert it.

In my case, I have an backend which sends me snakecase JSON and I need to convert it for the Vue3 framework. Everything worked fine until the backend send objects with numbers in it like:

```javascript
{
  address_line1: "Somestreet 15",
  address_line2: "next to Müller",
}
```

it gets converted into

```javascript
{
  addressLine1: "Somestreet 15",
  addressLine2: "next to Müller",
}
```

but converting it back to update server database objects I get:

```javascript
{
  address_line_1: "Somestreet 15",
  address_line_2: "next to Müller",
}
```

which is not the same. :-(

I used deepSnakeKeys()

I did some double convert tests to check other cases:

```typescript
  console.log(snakeCase("addressLine1")); // address_line_1 (should be address_line1)
  console.log(snakeCase("addressLine_1")); // address_line_1 (should be address_line_1)
  console.log(camelCase("address_line1")); // addressLine1 (ok)
  console.log(camelCase("1address_line2")); // 1AddressLine2 (not sure, but feels like it should be 1addressLine2)
  console.log(camelCase("address_line_1")); // addressLine1 (should be addressLine_1)
  console.log(snakeCase(camelCase("address_line2"))); // address_line_2 (should be address_line2)
  console.log(snakeCase(camelCase("address_line_2"))); // address_line_2 (ok)
  console.log(snakeCase(camelCase("1address_line2"))); // 1_address_line_2 (should be 1address_line2)
  console.log(camelCase(snakeCase("addressLine2"))); // addressLine2 (ok)
```

I used that library before and was about to switch over to here, but here are the same conversion errors. Because of that I do not know if I missed some camelCase to snake_case standards.

https://github.com/kbrabrand/snakify-ts/issues/6

**Comment by gustavoguichard** [comment]
Hi @RoNoLo , there's no right or wrong here. It is a matter of Style.

We chose to go with splitting numbers off words, for instance if you think in isolate, a function that turns `snakeCase('addressLine1')` into `"address_line1"` is making a weird assumption about style.
Why would the number go joined with the word?

We are also following the same pattern as lodash, which is a well known and accepted library, you can try the same there and will have the same result.

We actually discussed it already in #54 and here follows a bit of the reasoning I posted there, which was exactly all about consistency:
```
Approach 1:

snakeCase('World War 2') yields world_war_2
snakeCase(camelCase('World War 2')) also yields world_war_2

Approach 2:

snakeCase('World War 2') yields world_war_2
snakeCase(camelCase('World War 2')) yields world_war2
```

We chose approach 1 and I personally feel strongly that `address_line_1` reads much better than `address_line1`, and if we chose the latter we'd be disappointing the people that have chosen the first 🤷🏼 .

I'm going to close this issue too as I don't feel like there's a single solution for this case.

Best regards!

**Comment by RoNoLo** [comment]
Thanks for your great answer and explanation.

I just want to add a note here for clarification. I ONLY care about that issue because of object keys, because that is what I have to deal with. To be precise object keys in communication from server to frontends as JSON (via REST or similar formats), the thing is in PHP (in my case the Drupal Framework) underscore variable names are pretty common. In Vue3, the current frontend I use, camelCase variable names and therefore object keys are common or better required there. Thus converting them back and forth is needed. This is why I do have the strong need for object keys to stay consistent the same if I convert them back.

I would pretty much like if the string-ts function just had an argument on the deepSnakeKeys function to achieve that. I mean it's not about readability in my case, it's just, I have to deal with a backend (could be python, java, go, perl as well) which exposes JSON objects the way they do and require these objects to habe the same names when they sending them back. 

I could case less if "World War 2" is translated to "world_war_2", because that would be data. But as explained in my example I care about variable naming (which object keys essentially are), which should stay the same.

If you have an idea, the time or I can donate you an coffee/beer to add an argument to add a SnakeCase/camelCase version which keeps object key names the same it would be glorious. 

> Why would the number go joined with the word?
> Because it is not a captial letter.

Anyways, thanks for your time to make the developer world a better place.

**Comment by gustavoguichard** [comment]
@RoNoLo , I hear you!
Your use case was the exact use case I had which led me to create string-ts in the first place!

I used to deal with a lot of Rails APIs which are snake cased, then I created the [make-service](https://github.com/gustavoguichard/make-service) - which you should totally check out - and string-ts was born in there to make that exact conversion.
After working a lot on it I decided to extract it in its own package and further develop it.

The deal is AFAICT there's no way to save some "meta information" about the original string and make it go back to its original later so it ends up being a question of your API style. If you have control over it you can start naming your object keys differently but I'll assume that is not on the table for you.

Give me few hours as I'm just waking up and I'll post here a possible solution for you... hold on ☺️

**Comment by RoNoLo** [comment]
Good morning,

until 1 days ago, my understanding (because I care only about coding symbols), was that snake_case and camelCase are both notations which only care about letters, or better how the first letter of words is made more visible. It never crossed my mind that numbers (which are legal parts of coding language symbols) have a special treatment here.

thats why it struc me, that most snake to camel converters act different from what I thought.

For me it would go like this (and vis a vis):

addressLine -> address_line
addressLine100 -> address_line100
address_line_100 -> addressLine_100 // because its not a snaked letter
address_line_100_peter_parker -> addressLine_100PeterParker
1addressLine100 -> 1address_line100 // symbols with starting numbers are in most (?) languages forbidden, thus can be ignored

Basically: [A-Z] becomes _[a-z]

You know what I mean.

Anyways, my temporar solution is brute force .replace('foo_bar_1', 'foo_bar1') which is ofc terrible.

I hope to hear from you again and thank you for your time - so far.

**Comment by gustavoguichard** [comment]
Hi @RoNoLo !
In fact the way to work with casing functions is to split all the words in a string and change their cases accordingly.
The way we split the words is the matter of style... we chose to agree with lodash and most of the libraries in the JS ecosystem.

I hope you are not after this example you gave:  `address_line_100 -> addressLine_100`
Because that would be a whole different problem. But for the first requirements you had when you posted the issue I've got you.

I was able to put on a [draft PR](https://github.com/gustavoguichard/remix-global-loading-indicator/pull/2) in a repo I had so you can copy my process. It should help you on your project.

The secret here was to change the `string-ts` package according to your needs and then save the changes by using `patch-package`. If you lock `string-ts@1.3.2` in your package.json you shouldn't have problems going further.

Here follows the specific commit which shows you what to do: https://github.com/gustavoguichard/remix-global-loading-indicator/pull/2/commits/c1ceffce21157aa54552dae2df54a7286fcbd66a

I hope that helps you =)

**Comment by gustavoguichard** [comment]
@RoNoLo were you able to test it? Did it work?

**Comment by RoNoLo** [comment]
I am currently on a small holiday trip with the family. I'll report to you asap.
